scripts/api_diff_report/icore_module.py

```diff
@@ -0,0 +1,201 @@
+    module_info()
+
+
+def detect_changed_modules(changed_api_files):
+    """Detect changed modules based on changed API files."""
+    all_modules = module_info()
+    changed_modules = {}
+    for file_path in changed_api_files:
+        for k, v in all_modules.items():
+            if v['root_dir'] and v['root_dir'] in file_path:
+                changed_modules[k] = v
+                break
+
+    logging.info(f'changed_modules:\n{json.dumps(changed_modules, indent=4)}')
+    return changed_modules
+
+
+def module_info():
+    """retrive moudle info in MODULE_LIST from `.podspecs`
+    The moudle info helps to build Jazzy
+    includes: module name, source_files, public_header_files, language, umbrella_header, framework_root
+    """
+    module_from_podspecs = module_info_from_podspecs()
+    module_list = {}
+    for k, v in module_from_podspecs.items():
+        if k in MODULE_LIST:
+            if k not in module_list:
+                module_list[k] = v
+                module_list[k]['language'] = OBJECTIVE_C if v.get(
+                    'public_header_files') else SWIFT
+                module_list[k]['scheme'] = get_scheme(k)
+                module_list[k]['umbrella_header'] = get_umbrella_header(
+                    k, v.get('public_header_files'))
+                module_list[k]['root_dir'] = get_root_dir(
+                    k, v.get('source_files'))
+
+    logging.info(f'all_module:\n{json.dumps(module_list, indent=4)}')
+    return module_list
+
+
+def get_scheme(module_name):
+    """Jazzy documentation Info SWIFT only.
+
+    Get scheme from module name in .podspecs Assume the scheme is the
+    same as the module name:
+    """
+    MODULE_SCHEME_PATCH = {
+        'FirebaseInAppMessagingSwift': 'FirebaseInAppMessagingSwift-Beta',
+    }
+    if module_name in MODULE_SCHEME_PATCH:
+        return MODULE_SCHEME_PATCH[module_name]
+    return module_name
+
+
+def get_umbrella_header(module_name, public_header_files):
+    """Jazzy documentation Info OBJC only Get umbrella_header from
+    public_header_files in .podspecs Assume the umbrella_header is with the
+    format:
+
+    {module_name}/Sources/Public/{module_name}/{module_name}.h
+    """
+    if public_header_files:
+        if isinstance(public_header_files, list):
+            return public_header_files[0].replace('*', module_name)
+        elif isinstance(public_header_files, str):
+            return public_header_files.replace('*', module_name)
+    return ''
+
+
+def get_root_dir(module_name, source_files):
+    """Get source code root_dir from source_files in .podspecs Assume the
+    root_dir is with the format:
+
+    {module_name}/Sources or {module_name}/Source
+    """
+    MODULE_ROOT_PATCH = {
+        'FirebaseFirestore': 'Firestore/Source',
+        'FirebaseFirestoreSwift': 'Firestore/Swift/Source',
+        'FirebaseCrashlytics': 'Crashlytics/Crashlytics',
+        'FirebaseInAppMessagingSwift': 'FirebaseInAppMessaging/Swift/Source',
+    }
```

Comment:
Can't these be read from the podspec?

Reply:
Specifically, from the `spec.source_files` attribute.

Reply:
Yes and no. 
There are some inconsistency in the podspec. for example 
```
  base_dir = "FirebaseDatabase/Sources/"
  s.source_files = [
    base_dir + '**/*.[mh]',
    base_dir + 'third_party/Wrap-leveldb/APLevelDB.mm',
    base_dir + 'third_party/SocketRocket/fbase64.c',
    'FirebaseAuth/Interop/*.h',
    'FirebaseAppCheck/Interop/*.h',
    'FirebaseCore/Extension/*.h',
  ]
```
and
```
  source = 'FirebaseCombineSwift/Sources/'
  s.source_files = [
    source + '**/*.swift',
    source + '**/*.m',
  ]
```